File: src-tauri/src/schema/folder.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FolderCreateSchema {
    pub parent_id: Uuid,
    pub name: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FolderUpdateSchema {
    pub id: Uuid,
    pub parent_id: Option<Uuid>,
    pub name: Option<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FolderUpdateResponseSchema {
    pub id: Uuid,
    pub parent_changed: bool,
    pub name_changed: bool,
}

impl FolderUpdateResponseSchema {
    pub fn new(folder: &FolderUpdateSchema) -> Self {
        Self {
            id: folder.id,
            parent_changed: folder.parent_id.is_some(),
            name_changed: folder.name.is_some(),
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FolderNameCollisionSchema {
    pub is_unique: bool,
    pub colliding_folder: FolderResponseSchema,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FolderValidationSchema {
    pub id: Option<Uuid>,
    pub parent_id: Uuid,
    pub name: String,
    pub name_collision: Option<FolderNameCollisionSchema>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FolderResponseSchema {
    pub id: Uuid,
    pub parent_id: Uuid,
    pub name: String,
}

File: src/api/interface/folder.ts
import type { Id, IdentifiedObject } from "@/interface";

interface FolderProperties {
    parentId: Id;
    name: string;
}

export type FolderCreate = FolderProperties;

export interface FolderUpdate extends IdentifiedObject {
    parentId: Id | null;
    name: string | null;
}

export interface FolderBulkUpdateData {
    id: Id;
    parentChanged: boolean;
    nameChanged: boolean;
}

export interface FolderUpdateResponse extends FolderUpdate {
    parentChanged: boolean;
    nameChanged: boolean;
}

export type FolderResponse = IdentifiedObject & FolderProperties;

export interface FolderNameCollisionResponse {
    isUnique: boolean;
    collidingFolder: FolderResponse;
}

export interface FolderValidateResponse extends FolderProperties {
    id: Id | null;
    nameCollision: FolderNameCollisionResponse | null;
}
